File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i = 0;
        int j = 0;
        int maxlen = 0;

        HashSet<Character> set = new HashSet<>();

        while (i < s.length() && j < s.length()) {
            if (set.contains(s.charAt(j))) {
                maxlen = Math.max(maxlen, set.size());
                set.remove(s.charAt(i));
                i++;
            } else {
                set.add(s.charAt(j));
                j++;
            }
        }

       
        maxlen = Math.max(maxlen, set.size());

        return maxlen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        i = 0
        j = 0
        n = len(s)
        max = 0
        st = set()

        while j < n:
            if s[j] not in st:
                st.add(s[j])
                j += 1
            else:
                if j - i >= max:
                    max = j - i

                while i < j:
                    if s[i] == s[j]:
                        st.remove(s[i])
                        i += 1
                        break
                    else:
                        st.remove(s[i])
                        i += 1

                st.add(s[j])
                j += 1

        if j - i >= max:
            max = j - i

        return max
